[bitnami/opensearch] Enable automated setup of snapshot policies

lindhe opened the pull request:
### Description of the change

This change lets users opt-in to automatically initialize snapshot repositories and snapshot policies in OpenSearch using a post-install chart hook.

### Benefits

Before this change, configuring snapshot policies was a fairly complex task that requires interacting with the OpenSearch API manually. This change automates that process and allows for easy configuration via the Helm chart's values. I believe this will enable more people to have an OpenSearch instance that is properly configured so consistent backups can be taken.

### Possible drawbacks

Considering this is an opt-in feature, I don't see any significant draw-backs.

### Applicable issues

- fixes #29601

### Additional information

I've tried to split the PR into logically separate commits, so if we want the PVC creation to be a separate PR that's easy to do. Please inspect my commits for more info!

- good with objects for repo and policy?
- strange interdependence between persistance.enabled and snapshots.enabled

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

lindhe commented on bitnami/opensearch/README.md:
````diff
@@ -94,20 +94,19 @@ initScriptsSecret=special-scripts-sensitive
 
 ```yaml
-extraVolumes:
-  - name: snapshot-repository
-    nfs:
-      server: nfs.example.com # Please change this to your NFS server
-      path: /share1
-extraVolumeMounts:
-  - name: snapshot-repository
-    mountPath: /snapshots
+snapshots:
+  enabled: true
+  persistence:
+    enabled: true
 snapshotRepoPath: "/snapshots"
 ```
 
+For details, please refer to the `snapshots.*` documentation below.
+
````
Is this section of the README even relevant anymore? Should I remove it?

lindhe replied:
@jotamartos Do you think I can remove it?

jotamartos replied:
If we have a configuration for the snapshots, you can remove the usage of extraVolumes and extraVolumeMounts

lindhe replied:
I think I know what to do with the README as soon as we have landed in a decision regarding `snapshots.persistence.enabled` in the other thread. 👍

lindhe replied:
Considering we have decided to keep `false,false` as the defaults, I think think it's helpful to the users if we document that both needs to be set to true when enabling snapshots. Thus, I think we stay with my current changes like they are now.

lindhe commented on bitnami/opensearch/templates/_helpers.tpl:
```diff
@@ -822,3 +822,44 @@ Return true if a TLS credentials secret object should be created
+Create name for snapshot API repo data ConfigMap
+*/}}
+{{- define "opensearch.snapshots.repoDataConfigMap" -}}
+{{- printf "%s-repo-data" (include "opensearch.snapshots.fullname" $) -}}
```
Here (and in the matching helper for the policy cm name) I'm skipping the truncation to 63 characters. If the truncation leads to the suffix being dropped, we will have a resource collision, so I thought this was safer. I believe that the config map name is not limited to 63 characters, so it should be safe? What do you think?

lindhe commented on bitnami/opensearch/values.yaml:
```diff
@@ -3408,3 +3408,181 @@ dashboards:
+  ## NOTE: The field `settings.location` in each repo must match the snapshot repo path configured in `snapshotRepoPath`, if its `type` is `fs`.
+  ##
+  repositories:
+    default:
+      settings:
+        location: /snapshots # This must match the snapshot repo path configured in `snapshotRepoPath`
+      type: fs
```
I'd like some input on this design decision. Is it good to have the API body object as YAML like this? On the one hand, I like that everything is YAML so it's easy to work with. On the other hand, this kind of hides the fact that these are not "regular" values but actually YAML representations of OpenSearch API payloads. Might be clearer if we include them as JSON strings, but I think that would make it easier for users to introduce typos in their configs.

WDYT?

jotamartos replied:
As this is a yaml file, I think this configuration format is correct but you can use any other format here and change the template as required.

lindhe replied:
Certainly. I was just wondering what would provide the best user experience for the end user/consumer of the chart.

It also makes a difference for how we document the parameters. I think it's bad if the `@param` for this object becomes a copy-paste of the OpenSearch documentation. That will be a nightmare for you to maintain. If we just document this as a opaque blob, it's easier to say "this object must be an API object according to the OpenSearch documentation" and then they can look up the examples there.

I'll try and make it a JSON implementation to see if that looks better. I'll get back once I've updated my subjective opinion on the matter.

lindhe replied:
I just realized that my sanity checks (in [policies-cm.yaml](https://github.com/bitnami/charts/blob/c29ade4c05217605be43471a1338d1352c68af2c/bitnami/opensearch/templates/snapshots/policies-cm.yaml#L20) and [repos-cm.yaml](https://github.com/bitnami/charts/blob/c29ade4c05217605be43471a1338d1352c68af2c/bitnami/opensearch/templates/snapshots/repos-cm.yaml#L19-L24)) are entirely dependent on parsing the YAML. I think they are valuable to have, and it would be very hard to implement them unless we keep these objects as YAML code.

Thus, I'm marking this as resolved. If you disagree, please unresolve!

lindhe commented on bitnami/opensearch/values.yaml:
```diff
@@ -3408,3 +3408,181 @@ dashboards:
+    dataSource: {}
+    ## @param snapshots.persistence.enabled Enable persistence using Persistent Volume Claims
+    ##
+    enabled: false
```
I'm not convinced that it makes sense to let users disable PVC for snapshots. What kind of use-case is that? I think it is a strange interdependence between `snapshots.enabled` and `snapshots.persistence.enabled` right now; only `snapshots.enabled` should be required, in my opinion. But keeping this makes it more similar to other config in the chart.

WDYT?

jotamartos replied:
I prefer keeping the "snapshots.persistence.enabled" parameter and set it to true by default to follow the same approach we have for other solutions. For example, we mount an emptyVolume if persistence is not configured in MySQL

```
  {{- if and .Values.primary.persistence.enabled .Values.primary.persistence.existingClaim }}
        - name: data
          persistentVolumeClaim:
            claimName: {{ tpl .Values.primary.persistence.existingClaim . }}
  {{- else if not .Values.primary.persistence.enabled }}
        - name: data
          emptyDir: {}
...
```

You can follow that approach here. This case covers the scenarios where you can't create/configure a volume and need to disable persistence.

lindhe replied:
> This case covers the scenarios where you can't create/configure a volume and need to disable persistence.

I don't see any value for them to have `snapshots.enabled=true` unless they also have persistent volumes. But whatever – I'm happy to do what you suggest here and keep the config option as already configured.

jotamartos replied:
>Some components have persistence.enabled=true by default, but not all. At least dashboards.persistence.enabled=false by default. So there is no common approach to follow here.

As you are configuring an "extra" feature, I'd disable the persistence by default. Users can always enable it when deploying the solution

>We must not set snapshots.persistence.enabled=true by default unless we also set snapshots.enabled=true by default also. Otherwise, we'll just mount volumes that are never used, which makes no sense.

That's correct. 

>I suggest we remove the snapshots.persistence.enabled

I don't like this idea. I prefer providing the users with the possibility to manage this information. 

> or snapshots.enabled=false,snapshots.persistence.enabled=false by default.

This is the approach I'd follow. Sounds good?

lindhe replied:
Sounds all good. Thanks for the input. 👍

jotamartos commented:
Sorry for the delay in getting back to you. This is just a quick message to let you know that we are going to review it and will provide feedback as soon as possible.

lindhe commented:
Thanks, I really appreciate the ping to let me know you will get to it soon! 👍

jotamartos commented on bitnami/opensearch/values.yaml:
```diff
@@ -3408,3 +3408,181 @@ dashboards:
+  repositories:
+    default:
+      settings:
+        location: /snapshots # This must match the snapshot repo path configured in `snapshotRepoPath`
```
```suggestion
        location: /snapshots
```

As you added a note above, I'd remove the comment here

lindhe replied:
Yeah, I agree, it feels a bit redundant. Let's remove it and also let's make sure the README generator works with my new parameters.

jotamartos reviewed (commented):
I just added some comments. Please take a look at them and let me know so we finish reviewing this PR. Thanks for your contribution

jotamartos commented:
Please take a look at the error while generating the README ```error=Please ensure all *.registry params include the [default: REGISTRY_NAME] modifier in the chart bitnami/opensearch/values.yaml file
```

lindhe commented:
> Please take a look at the error while generating the README
> 
> ```
> error=Please ensure all *.registry params include the [default: REGISTRY_NAME] modifier in the chart bitnami/opensearch/values.yaml file
> ```

I'm aware of the error. It relates to this discussion: https://github.com/bitnami/charts/pull/29796#discussion_r1819152621

If we want to go for keeping the API body as YAML format, do you have any idea how we should go about configuring the `@param` argument for the documentation?

jotamartos commented:
Just updated the thread above. Please ping me with a new comment when you update the PR. Thanks

lindhe commented:
Great. I'll just run one last test by installing it in my local cluster to try and verify nothing is obviously broken. I'll ping you once I'm done with that.

lindhe commented on bitnami/opensearch/templates/snapshots/init-job.yaml:
```diff
@@ -0,0 +1,89 @@
+          {{- if .Values.snapshots.containerSecurityContext.enabled }}
+          securityContext: {{- include "common.compatibility.renderSecurityContext" (dict "secContext" .Values.snapshots.containerSecurityContext "context" $) | nindent 12 }}
+          {{- end }}
+          {{- if .Values.snapshots.resources }}
+          resources: {{- toYaml .Values.snapshots.resources | nindent 12 }}
+          {{- else if ne .Values.snapshots.resourcesPreset "none" }}
+          resources: {{- include "common.resources.preset" (dict "type" .Values.snapshots.resourcesPreset) | nindent 10 }}
+          {{- end }}
```
Seems like something's wrong with the resources. I'll check what's up.

lindhe replied:
Was a typo in nindent. Should be 12, not 10. Pushing fix in a minute.

lindhe commented:
I found a typo in the Network Policy that I fixed. When debugging this, I also added a flag to curl to make debugging things a lot easier when curl crashes.

I've tested this now and it seems to work! Assuming you approve my recent changes, I think we are ready to merge! 🚀

lindhe commented:
Hey @jotamartos , if you have time to check on my latest changes that would be great. 🙂

lindhe commented:
After merging main I'm getting a CI error I don't believe I got before:

![image](https://github.com/user-attachments/assets/74e8d897-b01b-492d-840c-77e4194baa81)

Is this a new check? I have no idea where to find the images warning list or why it's wrong.

lindhe commented:
@jotamartos I've resolved the conflicts now. If you could have a look at the latest changes and review before the next conflict emerge, that would be awesome! 👍

jotamartos commented on bitnami/opensearch/values.yaml:
```diff
@@ -3405,3 +3405,181 @@ dashboards:
+
+## @section OpenSearch Snapshots Parameters
+snapshots:
+  ## @param snapshots.command Override default container command (useful when using custom images)
+  ##
+  command:
+    - /usr/bin/curl
```
Is it necessary to add this here? I was thinking about writing the command directly in the template file

lindhe replied:
Not necessary at all, having it hard-coded in the template should work just as well for our purposes. Just that this lets the user customize the command, should they need to. We already let them change image, and if the do that it would very likely be required to change the command also (e.g. because `curl` exists in another path in their custom image).

There are `command` overrides in other parts of the `values.yaml` file too, the only difference here is that I included the default value as part of the `values.yaml` file instead of using logic in the template to check for custom values and then conditionally changing the command.

No rights or wrongs here, really. Just preference. But I do think that we need to accommodate for custom `command` if we want to let users provide custom images.

What do you think?

jotamartos commented on bitnami/opensearch/values.yaml:
```diff
@@ -3405,3 +3405,181 @@ dashboards:
+      drop: ["ALL"]
+    seccompProfile:
+      type: "RuntimeDefault"
+  ## @param snapshots.enabled Enable automatic setup of repositories and snapshot policies
+  ##
+  enabled: false
```
I'd move this to the beginning of the section to give more visibility

lindhe replied:
Sure. I sorted it alphabetically, but visibility makes sense. I'll move it.

lindhe replied:
Should `snapshots.persistence.enabled` be moved to the top of `snpashots.persistence`, so they are similar?

lindhe replied:
I've moved both to the top. If you don't want the same for persistence, I'll just revert the commit for you. Please mark this conversation as resolved if you are happy with the solution!

jotamartos reviewed (commented):
Two minor comments. Please have a look
